libminifi/include/utils/net/SessionHandlingServer.h

```diff
@@ -0,0 +1,67 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <utility>
+#include <memory>
+
+#include "Server.h"
+#include "asio/ssl.hpp"
+
+namespace org::apache::nifi::minifi::utils::net {
+
+template<typename SessionType>
+class SessionHandlingServer : public Server {
+ public:
+  SessionHandlingServer(std::optional<size_t> max_queue_size, uint16_t port, std::shared_ptr<core::logging::Logger> logger)
+      : Server(max_queue_size, std::move(logger)),
+        acceptor_(io_context_, asio::ip::tcp::endpoint(asio::ip::tcp::v4(), port)) {
+  }
+
+  void run() override {
+    startAccept();
+    Server::run();
+  }
+
+ protected:
+  void startAccept() {
+    auto new_session = createSession();
+    acceptor_.async_accept(new_session->getSocket(),
+                           [this, new_session](const auto& error_code) -> void {
+                             handleAccept(new_session, error_code);
+                           });
```

Comment:
Sounds logical, but for some reason changing it to unique_ptr causes a segmentation fault in the asio library. I wasn't able to figure out what caused it so far.

Reply:
Did you query the socket in a separate statement? Because the evaluation order of function arguments are unspecified, so if you don't extract the socket into a separate variable, you can easily move from new_session (as part of the lambda expression) before calling getSocket on it.

Reply:
You are right, that caused this issue. Unfortunately there is another problem with not using `shared_ptr` here. If we use `unique_ptr` instead `TcpSession::start` fails with `bad_weak_ptr` when it tries to use `shared_from_this()`.

Reply:
ok, let's leave it then